src/Mod/Sketcher/App/SketchAnalysis.cpp

```diff
@@ -222,67 +233,120 @@ int SketchAnalysis::detectMissingPointOnPointConstraints(double precision, bool
+
+    for (auto &constraint:sketch->Constraints.getValues()) {
+        if (constraint->Type == Sketcher::Coincident ||
+            constraint->Type == Sketcher::Tangent ||
+            constraint->Type == Sketcher::Perpendicular) {
+            coincidences.push_back(constraint);
+        }
+        // TODO optimizing by removing constraints not applying on vertices ?
     }
 
-    std::sort(vertexIds.begin(), vertexIds.end(), Vertex_Less(precision));
+    std::list<ConstraintIds> missingCoincidences; //Holds the list of missing coincidences
+
     std::vector<VertexIds>::iterator vt = vertexIds.begin();
     Vertex_EqualTo pred(precision);
 
-    std::list<ConstraintIds> coincidences;
-    // Make a list of constraint we expect for coincident vertexes
+    // Comparing existing constraints and find missing ones
+
     while (vt < vertexIds.end()) {
-        // get first item whose adjacent element has the same vertex coordinates
+        // Seeking for adjacent group of vertices
         vt = std::adjacent_find(vt, vertexIds.end(), pred);
-        if (vt < vertexIds.end()) {
+        if (vt < vertexIds.end()) { // If one found
             std::vector<VertexIds>::iterator vn;
-            for (vn = vt+1; vn != vertexIds.end(); ++vn) {
+            std::set<VertexIds, VertexID_Less> vertexGrp; // Holds a single group of adjacent vertices
+            // Extract the group of adjacent vertices
+            vertexGrp.insert(*vt);
+            for (vn = vt+1; vn < vertexIds.end(); ++vn) {
                 if (pred(*vt,*vn)) {
-                    ConstraintIds id;
-                    id.Type = Coincident; // default point on point restriction
-                    id.v = vt->v;
-                    id.First = vt->GeoId;
-                    id.FirstPos = vt->PosId;
-                    id.Second = vn->GeoId;
-                    id.SecondPos = vn->PosId;
-                    coincidences.push_back(id);
+                    vertexGrp.insert(*vn);
                 }
                 else {
                     break;
                 }
             }
 
-            vt = vn;
-        }
-    }
+            std::vector<std::set<VertexIds, VertexID_Less>> coincVertexGrps; // Holds groups of coincident vertices
+
+            // Decompose the group of adjacent vertices into groups of coincident vertices
+            for (auto &coincidence:coincidences) { // Going through existent coincidences
+                VertexIds v1, v2;
+                v1.GeoId = coincidence->First;
+                v1.PosId = coincidence->FirstPos;
+                v2.GeoId = coincidence->Second;
+                v2.PosId = coincidence->SecondPos;
+
+                // Look if coincident vertices are in the group of adjacent ones we are processing
+                auto nv1 = vertexGrp.extract(v1);
+                auto nv2 = vertexGrp.extract(v2);
+                if (nv1.empty() && nv2.empty()) continue;
+
```

Comment:
At this point nv1 or nv2 could be empty but not both

Reply:
Actually both can be empty. ;) I have a test file for that. But the problem is that continuing isn't the right solution in all cases. :| I have to polish this PR.

Reply:
Actually, when both empty, in no cases it is good idea to continue. We should always check. It's fixed now.

Reply:
> Actually both can be empty. ;)

Sure, but in your first version you had a `continue;` in the loop when nv1 and nv2 were empty. So, after that line only nv1 or nv2 could be empty but not both and this fact wasn't checked in the further code.